### PR TITLE
Fix bias update in scoped all reduce

### PR DESCRIPTION
**Pull request by skavulya**
# What does this PR do?
Fixes the bias update in `post_all_reduce` for scoped all reduce to remove assumption of in-place bias updates. This led to incorrect outputs with fp8 models like starcoder2 since the Intel Neural Compressor does not use in-place updates (see https://github.com/intel/neural-compressor/blob/27f3e2657b2667e8bca8fb9c02a50d55f404a7e6/neural_compressor/torch/algorithms/fp8_quant/_quant_common/helper_modules.py#L417)

This PR is a more general fix for the issue than https://github.com/huggingface/optimum-habana/pull/1432 which only fixed the issue for bf16 models.

`QUANT_CONFIG=./quantization_config/maxabs_measure.json python ../gaudi_spawn.py --use_deepspeed --world_size 2 \
run_generation.py --model_name_or_path bigcode/starcoder2-15b \
--use_hpu_graphs --trust_remote_code --attn_softmax_bf16 --trim_logits \
--use_kv_cache --use_flash_attention --flash_attention_recompute --max_new_tokens 128 \
--batch_size 1 --bf16 --prompt "def is_prime():"`

**Output before fix:**
output 1: ('def is_prime():\x89tFo_Fo_ \\s%nu|\\\\LL_LL)\ntt=> namemap (LL-\\\\LL0LL-LL-\\\\LL testInstances",LLÏ????LL testInstances",LL toolSettings:LL;LL-\\\\LLLLjooFoLL-LLJJLL-LLLLLL-\\\\LLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLL toolSettings (LL namemap \\LLLLLL toolSettings (LLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLL namemap \\LL_LLLLLLLLLLLL',)

**Output after fix**
output 1: ('def is_prime():\n    for i in range(2, int(math.sqrt(n)) + 1):\n        if n % i == 0:\n            return False\n    return True\n\ndef is_palindrome():\n    return str(n) == str(n)[::-1]\n\ndef is_lychrel():\n    for i in range(50):\n        n = n + int(str(n)[::-1])\n        if is_palindrome():\n            return False\n    return True\n\ncount = 0\nfor i in range(10000):\n    if is_lychrel():',)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yafshar** [comment]
@skavulya the 1.14 release is already done https://github.com/huggingface/optimum-habana/releases/tag/v1.14.0
Probably this fix can be added to any upcoming patch release. I will start reviewing that

**Comment by yafshar** [comment]
@skavulya I agree, this is a better fix than https://github.com/huggingface/optimum-habana/pull/1432. Can you please add the test results before and after this PR.

**Comment by skavulya** [comment]
Thanks @yafshar I'll run the tests and include the results

File: optimum/habana/transformers/models/starcoder2/modeling_starcoder2.py [hunk 1]
```diff
@@ -307,7 +308,8 @@ def pre_attn_forward(
 
             if q_len == 1:
                 # next token
-                with ht.sdp_kernel(enable_recompute=False):
+                use_recompute = True if os.getenv("QUANT_CONFIG", "") else False
```
Comment (yafshar) [inline optimum/habana/transformers/models/starcoder2/modeling_starcoder2.py:311]:
```suggestion
                use_recompute = bool(os.getenv("QUANT_CONFIG"))
```

personal suggestion, less code!

**Comment by skavulya** [comment]
The transformer tests yield the same results as main.

`GAUDI2_CI=1 RUN_SLOW=true python -m pytest tests/transformers/tests/models/`

main: commit  03fa6dd50438549c
==== 41 failed, 993 passed, 401 skipped, 101 warnings in 1404.02s (0:23:24) ====

this pr:
==== 41 failed, 993 passed, 401 skipped, 100 warnings in 1629.37s (0:27:09) ====

**Review by yafshar — APPROVED** [review]
LGTM!

@regisss, this PR is ready. Would you please check it.

**Comment by jiminha** [comment]
@xin3he  Can you check why neural compressor is not using in-place bias updates here?  Could you also tag Eran here as well?

**Comment by xin3he** [comment]
@ulivne Could you please help have a look here?

**Review by Yantom1 — APPROVED** [review]

**Review by MrGeva — APPROVED** [review]
LGTM

**Review by tthakkal — APPROVED** [review]

**Review by jiminha — APPROVED** [review]

**Pull request merged**